x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java

```diff
@@ -6,54 +6,80 @@
 import java.util.Map;
 import java.util.Set;
 
 import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.xpack.core.security.action.UpdateIndexMigrationVersionAction.MIGRATION_VERSION_CUSTOM_DATA_KEY;
+import static org.elasticsearch.xpack.core.security.action.UpdateIndexMigrationVersionAction.MIGRATION_VERSION_CUSTOM_KEY;
+import static org.elasticsearch.xpack.core.security.test.TestRestrictedIndices.INTERNAL_SECURITY_MAIN_INDEX_7;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
 
 public class SecurityIndexRolesMetadataMigrationIT extends AbstractUpgradeTestCase {
 
-    public void testMetadataMigratedAfterUpgrade() throws Exception {
-        String testRole = "test-role";
-        String metaKey = "test_key";
-        String metaValue = "test_value";
-
-        Map<String, Object> testMetadata = Map.of(metaKey, metaValue);
+    public void testRoleMigration() throws Exception {
+        String oldTestRole = "old-test-role";
+        String oldMetaKey = "old-meta-test-key";
+        String oldMetaValue = "old-meta-test-value";
+        String mixed1TestRole = "mixed1-test-role";
+        String mixed1MetaKey = "mixed1-meta-test-key";
+        String mixed1MetaValue = "mixed1-meta-test-value";
+        String mixed2TestRole = "mixed2-test-role";
+        String mixed2MetaKey = "mixed2-meta-test-key";
+        String mixed2MetaValue = "mixed2-meta-test-value";
+        String upgradedTestRole = "upgraded-test-role";
+        String upgradedMetaKey = "upgraded-meta-test-key";
+        String upgradedMetaValue = "upgraded-meta-test-value";
         if (CLUSTER_TYPE == ClusterType.OLD) {
-            createRole(testRole, testMetadata);
-            assertEntityInSecurityIndex(testRole);
-        }
-        if (CLUSTER_TYPE == ClusterType.UPGRADED) {
-            refreshSecurityIndex();
-            waitForMigrationCompletion();
-            assertEntityInSecurityIndex(testRole, metaKey, metaValue);
-        }
-    }
-
-    public void testMetadataWrittenAfterUpgradeWithoutMigration() throws IOException {
-        String testRole = "another-test-role";
-        String metaKey = "another-test_key";
-        String metaValue = "another-test_value";
-
-        Map<String, Object> testMetadata = Map.of(metaKey, metaValue);
-
-        if (CLUSTER_TYPE == ClusterType.UPGRADED) {
-            createRole(testRole, testMetadata);
-            assertEntityInSecurityIndex(testRole, metaKey, metaValue);
+            createRoleWithMetadata(oldTestRole, Map.of(oldMetaKey, oldMetaValue, "meta", "test"));
+            assertDocInSecurityIndex(oldTestRole);
+            if (canRolesBeMigrated() == false) {
+                assertNoMigration(adminClient());
+                assertCannotQueryRolesByMetadata(client());
+            }
+        } else if (CLUSTER_TYPE == ClusterType.MIXED) {
+            if (FIRST_MIXED_ROUND) {
```

Comment:
What does this do? Can you add a comment explaining?

Reply:
Ah thanks for clarifying! For the things we're testing, does it make sense to cover that case? I'll leave it up to you but it feels like maybe it doesn't impact our implementation if one or two nodes are upgraded. Always good with some extra coverage though so I don't feel too strongly.

Reply:
>I'll leave it up to you but it feels like maybe it doesn't impact our implementation if one or two nodes are upgraded

Yeah, it's not crucial.
The issue that prompted me to change it was that the role, in a mixed cluster, was always double-created (which was odd).